### DISCUSSION TITLE
Is there a way to set common AsyncLocal for all tests in a test fixture?

### DISCUSSION THREAD

**Discussion by alewmt** [nunit/nunit#3991] [Q&A] [answered]
I have to set a context with AsyncLocal inside in every test:
```
[Test]
public async Task Test()
{
    using(MyAsyncLocalContext.Create())
    {
        //test body
    }
}

```
I expect that where is some common root for all test in a fixture and it will be very convenient to set context only once. 
Something like this
```
[SetUp]
public async Task Setup()
{
    CommonRoot.Add((fixtureCall) =>
    {
        using(MyAsyncLocalContext.Create())
        {
            fixtureCall()
        }
    )
}
```

**Comment by jnm2** [comment, marked as answer]
I think so, using an attribute that implements `IWrapSetUpTearDown`. The delegating command would wrap the `innerCommand.Execute` call with `using (MyAsyncLocalContext.Create())`. You could apply the attribute to the test fixture, a test method, or the whole assembly.

**Reply by alewmt** [reply]
It works with test method only.

**Reply by jnm2** [reply]
Ah, you're right. It would be `ITestAction` if you want to do it for everything in a fixture or assembly: https://docs.nunit.org/articles/nunit/extending-nunit/Action-Attributes.html

**Reply by jnm2** [reply]
Though without trying it, I'm not sure if AsyncLocal will work with Before/After method pairs.
